src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java

```diff
@@ -41,7 +41,6 @@ public UploadLettersTask(
         this.availabilityChecker = availabilityChecker;
     }
 
-    @Transactional
```

Comment:
Wouldn't this be easier if we had moved this to `uploadLetter` and make that method public?

Reply:
If all the letters are marked Uploaded in the same transaction then they either all be committed or all rolled back; having a second transaction won't change this.

Each letter needs to be a separate transaction which is why streams don't work, with streams the order of events is as follows:

BEGIN TRANSACTION
for each letter in stream {
  upload letter
  UPDATE LETTER to Uploaded
}
COMMIT
Nothing will be committed until the iteration completes and the outer transaction commits, and if anything goes wrong the transaction is rolled back.

I'm fairly sure the only way around this is to break this up into multiple transactions. See the [last comment here](https://jira.spring.io/browse/DATAJPA-1023)

Reply:
Though we might be talking about different things; I still need to fix the cause of the SQLException in sandbox, that requires a separate investigation.

Reply:
Still, the streams approach is not going to work if we have 100,000 letters to upload.